### ISSUE TITLE
Control retry time within job

### ISSUE THREAD

**Issue by sezaru** [koudelka/honeydew#103]
Hello, I'm using honeydew to send HTTP requests to a third-party endpoint. This endpoint request can fail for some reasons and their recommendation is to do an Exponential Backoff, which I get out-of-box with your great library.

Another possibility is that they can send me a Retry-After header with the time I need to wait before retrying.

Since I will only know that value inside the job itself, is there any wait to, if I know the time I need to wait, use that time as the retry delay instead of the Exponential Backoff?

Basically this means that I would do the request, if fail, I would parse it anyway and get the Retry-After header, if it is present, I want to delay the next retry with the time they sent me, if it is not present or the job crashed for some other reason, I would fallback to Exponential Backoff

**Comment by koudelka** [comment]
Hey @sezaru,

That's not something that's explicitly supported at the moment, but if I'm understanding my own code correctly, you should be able to communicate a value from within the job to the failure mode by using a throw.

In the job, you'd `throw {:retry_secs, 10}`.

Then you'd implement a custom failure mode something like this:

```elixir
defmodule Honeydew.FailureMode.JobDirectedRetry do
  alias Honeydew.Job
  alias Honeydew.FailureMode.Retry
  alias Honeydew.Crash

  @behaviour Honeydew.FailureMode

  @impl true
  def validate_args!(args), do: Retry.validate_args!(args, __MODULE__)

  @impl true
  def handle_failure(job, reason, args) do
    args = Keyword.put(args, :fun, &secs_from_throw/3)

    Retry.handle_failure(job, reason, args)
  end

  def secs_from_crash(%Job{failure_private: nil} = job, reason, args) do
    exponential(%Job{job | failure_private: 0}, reason, args)
  end

  def secs_from_throw(
    %Job{failure_private: times_retried},
    %Crash{type: :throw, reason: {:retry_secs, retry_secs}},
    %{times: max_retries}
  ) when times_retried < max_retries do
    {:cont, times_retried + 1, retry_secs}
  end

  def secs_from_throw(_, _, _), do: :halt
end
```

I haven't tested the above, but it should be enough to get you started in the right direction. :)

**Comment by sezaru** [comment]
It sure is! Thank you for the help, now it is clear to me how to achieve that.
